### PR TITLE
add trusted setup ceremony explainer

### PR DESCRIPTION
**Pull request by dcbuild3r**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,12 @@
...
+[World ID](https://worldcoin.org/blog/announcements/introducing-world-id-and-sdk) is a privacy-preserving proof of personhood protocol that leverages the [Semaphore protocol[(https://semaphore.appliedzkp.org/)] in order to prove inclusion of members in a merkle tree where each identity commitment was generated by an orb and its corresponding private key is stored on our users' phones. In the Semaphore protocol implementation there are individual sequential insertions. Individual insertions for big merkle trees are very expensive and therefore economically unfeasible at the World ID userbase scale (current numbers can be found [here](https://worldcoin.org/)). In order to resolve that issue we developed custom circuits written in gnark in order to do batch insertions into the Semaphore merkle tree. This gnark circuit leverages the groth16 proof system on the bn254 curve and requires a custom trusted setup ceremony to be made in order to achieve verifier soundness.
```
Comment (recmo) [inline README.md]: Can we change WorldID link to something up-to-date and maintained? Like whitepaper or docs?

File: README.md [hunk 2]
```diff
@@ -1,6 +1,12 @@
 # Worldcoin trusted setup ceremony
```
Comment (recmo) [inline README.md]: "Worldcoin" is too generic. "Semaphore Merkle Tree Batcher" is more specific.

File: README.md (hunk 1 above)
Comment (recmo) [reply README.md]: Actually this whole paragraph is not very relevant. We should be pointing to the "Semaphore Merkle Tree Batcher".

That we use it for WorldID is an implementation detail. Maybe worth mentioning but not very relevant.

**Review by recmo — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (dcbuild3r) [reply README.md]: Will change

File: README.md (hunk 1 above)
Comment (dcbuild3r) [reply README.md]: Will change

File: README.md [hunk 3]
```diff
@@ -1,10 +1,16 @@
...
...rldcoin.org/proof-of-personhood) is a privacy-preserving proof of personhood protocol that leverages the [Semaphore protocol](https://semaphore.appliedzkp.org/) in order to prove inclusion of members in a merkle tree where each leaf was verified using some method (e.g. [the orb](https://whitepaper.worldcoin.org/technical-implementation)), inserted into the tree and where its corresponding private key is self-custodied by each member. In the Semaphore protocol implementation there are individual sequential insertions. Individual insertions for big merkle trees are very expensive and therefore economically unfeasible onchain operations at the World ID userbase scale (user count can be found [here](https://worldcoin.org/)). In order to scale our efforts we developed the [Semaphore Merkle Tree Batcher](https://github.com/worldcoin/semaphore-mtb) (SMTB) which is composed of custom zero-knowledge circuits written in [gnark](https://github.com/ConsenSys/gnark) in order to do batch insertions into the Semaphore merkle tree. This gnark circuit leverages the groth16 proof system on the bn254 curve and requires a custom trusted setup ceremony to be made in order to achieve verifier soundness.
```
Comment (recmo) [inline README.md]: Still way to much emphasis on the mostly irrelevant point that it is used by WorldID. How about:

---

This repo contains the transcript of trusted setup ceremony for [Semaphore Merkle Tree Batcher](https://github.com/worldcoin/semaphore-mtb). Please see that repo for details.

To verify the transcript do

```
<commands>
```

to generate prover and verifier keys do

```
commands
```

To learn more about trusted setup ceremonies read the [BGM17] paper or the explainer by X [link]

File: README.md [hunk 4]
```diff
@@ -390,4 +477,51 @@ And the last step before deploying to production is to generate Solidity verifie
...
 
+### Production
+
+Deployed [`semaphore-mtb`](https://github.com/worldcoin/semaphore-mtb/) verifier contracts and their corresponding verifying keys can be found here:
+
+- Batch size 10, tree depth 30: [Etherscan](https://etherscan.io/address/0x6e37bAB9d23bd8Bdb42b773C58ae43C6De43A590#code)
+- Batch size 100, tree depth 30: [Etherscan](https://etherscan.io/address/0x03ad26786469c1F12595B0309d151FE928db6c4D#code)
+- Batch size 1000, tree depth 30: [Etherscan](https://etherscan.io/address/0xf07d3efadD82A1F0b4C5Cc3476806d9a170147Ba#code)
+
+### Verifying the ceremony
```
Comment (recmo) [inline README.md]: We probably want this much higher up, near the top. Or link to it from the top if you feel strongly that we need to keep a chronological ordering (I don't think we need to). Right now the readme starts with instructions on how to contribute, but nobody coming to this readme will be interested in that. Most likely they want to know what it is, how it works (both already well explained) and why they should trust it.

File: README.md [hunk 5]
```diff
@@ -390,4 +477,51 @@ And the last step before deploying to production is to generate Solidity verifie
...
+
+We can verify the keys generated in this ceremony are being used in the production contracts we see above by following the rest of this section.
+
```
Comment (recmo) [inline README.md]: Missing two important steps:

* Verify that you trust at least one of the participants. (And make it clear you don't need to trust all of them).
* Verify that they actually contributed and you didn't just write their names down. (i.e. link to the github comments)

**Review by recmo — APPROVED** [review]
LGTM. Can improve a little more.

File: README.md (hunk 4 above)
Comment (dcbuild3r) [reply README.md]: Understood

File: README.md (hunk 5 above)
Comment (dcbuild3r) [reply README.md]: Will explain. Thanks!

**Pull request merged**